### PR TITLE
Backport/metadata handling issue fixed in master

### LINKED ISSUE #19: Multiline filter does not handle @metadata properly
Hi,

I've been using <code>@metadata</code> in a configuration file for Logstash 1.5.4. My config is using metadata to represent the type of an event. I use metadata instead of type because I don't want the type field in my output (and I have multiple config in a single directory and each output will be different depending on the type). Anyway, here is a trimmed-down version of my config:

```
input {
    file {        
        path => "/tmp/web.log"        
        sincedb_path => "/dev/null"
        start_position => "beginning"
        add_field => { "[@metadata][type]" => "web" }
    }
}
filter {    
    multiline {                        
        pattern => "^\s"
        negate => false
        what => "previous"            
    }      
}
output {    
    if [@metadata][type] == "web" {
        stdout { codec => rubydebug { metadata => true } }        
    }
}
```
My test file contains 3 lines (with proper line ending):

```
line 1
line 2
line 3
```
When I execute this config, the last line is never output. If I remove the <code>if [@metadata][type] == "web"</code> test, that last line is output.

I looked at the code of the multiline plugin and found that metadata are not always handled properly (if I am not mistaken). In the case of the previous filter, when the event does not match the pattern, the current event is kept aside in a tmp variable and the pending events are all merged together to replace the current event. The current event, stored in tmp variable, is put in the pending array until the next event comes or a flush is triggered. Problem is, when the current event is stored in tmp, the metadata are not kept (because it is using .to_hash instead of .to_hash_with_metadata). Also, when the pending events are all merged together, they are passed to event.overwrite, which, AFAICT, is not handling metadata either. In case of a next filter, we have the same problem because of event.overwrite.

I've written a [small patch](https://github.com/pascaldimassimo/logstash-filter-multiline/commit/2b1f1a1df10b04215b9b63c6d935e7097b368a9a) that fixes the issue with my configs. I am not sure it cover all cases and if we should modify event.overwrite also. Let me know if you'd like me to make further changes to it. Or feel free to use the code for a fix if deemed necessary.

Thanks!

### PR DESCRIPTION
**Pull request by purbon**
Fixes #19  and #22 for the core-api-v1 (aka 1.x logstash-core) series of plugins.

### CODE REVIEW

**Comment by purbon** [comment]
@jsvd can you let me know your thought on this backport?

File: logstash-filter-multiline.gemspec [hunk 1]
```diff
@@ -20,7 +20,7 @@ Gem::Specification.new do |s|
   s.metadata = { "logstash_plugin" => "true", "logstash_group" => "filter" }
 
   # Gem dependencies
-  s.add_runtime_dependency "logstash-core", '>= 1.4.0', '< 2.0.0'
+  s.add_runtime_dependency "logstash-core", '>= 1.4.0', '< 1.999.0'
```
Comment (jsvd) [inline logstash-filter-multiline.gemspec]: maybe `s.add_runtime_dependency "logstash-core", '>= 1.4.0', '~> 1'` ?

File: Gemfile [hunk 2]
```diff
@@ -1,2 +1,3 @@
 source 'https://rubygems.org'
-gemspec
\ No newline at end of file
+gemspec
+gem "concurrent-ruby", "0.9.1"
```
Comment (jsvd) [inline Gemfile]: this doesn't seem necessary

File: logstash-filter-multiline.gemspec (hunk 1 above)
Comment (purbon) [inline logstash-filter-multiline.gemspec]: @jsvd It was agreed with @colinsurprenant this was the right way to follow.

File: Gemfile (hunk 2 above)
Comment (purbon) [inline Gemfile]: it is if you want to fix the developer experience.

File: logstash-filter-multiline.gemspec (hunk 1 above)
Comment (purbon) [inline logstash-filter-multiline.gemspec]: see https://github.com/elastic/logstash/issues/4236 for details.

File: Gemfile (hunk 2 above)
Comment (ph) [inline Gemfile]: 1.5.6 will ship with 0.9.2 and this will break.

Comment (purbon) [inline Gemfile]: @ph will help if I add a group there only for testing? I want the user who checkout this does not encounter this issue.

**Comment by jsvd** [comment]
Code LGTM, but we need to address the concurrent-ruby version problem before merging..
so if logstash-core is <=1.5.5 concurrent-ruby needs to be 0.9.1, but if > 1.5.5 then 0.9.2?

**Comment by purbon** [comment]
@jsvd as it's right now the gemfile is only helping on development environment. It will be harmless when installing the plugin/gem. what do you think?

**Comment by jsvd** [comment]
but doing bundle install && bundle exec rspec will break as soon as 1.5.6 is released, right?
maybe we can pin down the logstash-core version only in the group?

**Comment by purbon** [comment]
well I guess as soon as 1.5.6 gets released we don't need this anymore. what do you think?

**Comment by jsvd** [comment]
ahh because 1.5.6 pins a specific version of concurrent-ruby instead of ~0.9.1?
that still means we'll need to come to this plugin and remove the group definition, right?

**Comment by jsvd** [comment]
+1 on removing the concurrent-ruby pinning
otherwise LGTM

**Comment by jsvd** [comment]
ah and please change the logstash-core dependency to < 2.0.0.alpha0 ?

**Comment by purbon** [comment]
works for me, thanks for your feedback.

**Pull request merged**